### PR TITLE
feat: Add a planner rule that expands logical join nodes

### LINKED ISSUE #4627: Add a planner rule that expands logical join nodes
For the sake of performance, it is important that the new `join` function receives sorted tables as its input.

To ensure this is the case, we should add a planner rule that finds logical `join` nodes, and expands them into two physical nodes: one `sort` followed by a `sortMergeJoin`.

A separate but related issue has been filed to add another planner rule that removes unnecessary `sort` nodes: https://github.com/influxdata/flux/issues/4628

NOTE: Decide whether to change the existing join rule (re predicate function) or add a new rule

DOD:
- The planner rule described above exists, and is used by the planner

### PR DESCRIPTION
**Pull request by Marwes**
Just implements the pattern matching and tests that the rule is used. There isn't an actual
transformation that makes use of this.

Closes #4627

### CODE REVIEW

**Comment by Marwes** [comment]
@scbrickley Is this all that was needed for the issue? There isn't any way to make use of the planner rule so I were a bit confused about the scope of this.

File: stdlib/join/sort_merge_join.go
Comment (scbrickley) [inline stdlib/join/sort_merge_join.go:98]: I don't think there's any need to replace the node here. There's another rule in place that will take care of inserting the proper procedure spec (see #4665).

**Review by scbrickley — CHANGES_REQUESTED** [review]
This isn't quite what I had in mind, but I realize now that I failed to outline some important details in the issue.

Happy to discuss these points further.

File: stdlib/join/sort_merge_join.go
Comment (Marwes) [reply stdlib/join/sort_merge_join.go]: Not sure how I would stop the rewrite from just repeatedly being applied without doing a replacement (and thereby ensuring that the rewrite rule no longer matches).

File: stdlib/join/sortmergejoin_test.go
Comment (Marwes) [reply stdlib/join/sortmergejoin_test.go]: Yeah, I copy pasted this a bit to quickly. Think I have fixed it now, though I am not entirely sure how to check that predecessor 0 and 1 matches the left and right parameter so the sorts may be the wrong way around :thinking: .

File: stdlib/join/sort_merge_join.go
Comment (scbrickley) [inline stdlib/join/sort_merge_join.go:45]: Could we copy the `On` parameter here as well?

**Review by scbrickley — APPROVED** [review]
I think this looks good. I had one comment, but it's not worth blocking this PR over it.

I share your concerns about knowing which parent node is which, but because there's no actual transformation running yet, I don't think that's a big deal. I'm going to rebase my current branch for #4629 (which contains the actual transformation code) on this once it merges. The tests I write for that PR should be enough to confirm this works as intended.

**Pull request merged**
